modules/grizzly/src/main/java/org/glassfish/grizzly/nio/NIOConnection.java

```diff
@@ -866,8 +857,7 @@ private void notifyCloseListeners(final CloseReason closeReason) {
         final org.glassfish.grizzly.CloseType closeType =
                 closeReason.getType();
         
-        org.glassfish.grizzly.CloseListener closeListener;
-        while ((closeListener = closeListeners.poll()) != null) {
+        for (org.glassfish.grizzly.CloseListener closeListener: closeListeners.keySet()) {
```

Comment:
closeListeners.poll() would remove the listeners from the queue.  Seems something similar should be done.

TBH, for 2.3.x, I can't imagine the add/remove methods would be a hot spot.  I'd be fine considering synchornization on the closeListeners instance when performing mutations and just using a HashSet.